In [4]:
import pandas as pd, numpy as np
import matplotlib.pyplot as plt
from datetime import datetime

%matplotlib inline

# Performance Analysis

In [6]:
root = "C:/Users/workspace/brad_public_workspace_on_win/SH_tongliang/strats/"
perf = pd.read_csv(root+"11_8_Ranking_Original_Half_Commissions_3_5_50_drawdown_analysis_csv.csv")
del perf['Unnamed: 0']

In [7]:
perf_ = perf.drop(["公式应用"], axis=1)

In [8]:
perf_.sort_values("商品").head()

类型    商品        建仓时间   建仓价格        平仓时间   平仓价格   数量    交易成本        净利  \
51   空头  DADA   2018-7-26  213.7    2018-8-6  188.5  467  234.78  11533.62   
133  空头  DADA   2018-9-27  131.4   2018-9-29  132.6  761  251.13  -1164.33   
31   多头  DADA   2018-7-24  215.5   2018-7-25  219.1  464  252.07   1418.33   
112  空头  DADA    2018-9-9  122.3   2018-9-17  101.4  817  228.45  16846.85   
182  空头  DADA  2018-10-21  118.7  2018-10-22  119.3  842  250.50   -755.69   

          累计净利         收益率       累计收益率  
51    40825.18   53971.06%   67381.98%  
133  193968.00   -8860.96%  330701.44%  
31     5599.41    6581.59%  -12656.34%  
112  104490.00  137750.17%  186721.57%  
182  183757.50   -6366.43%   93968.98%

In [9]:
perf_.columns = ['position_direction', 'asset', 'entry_date', 'entry_price', 'exit_date', 'exit_price', 'qty', 'trade_costs', 'net_gains', 'cum_gains',
       'returns', 'cum_returns']

In [10]:
perf_.head(3)

position_direction asset entry_date  entry_price exit_date  exit_price  qty  \
0                 空头  DXRP   2018-7-3        775.5  2018-7-4       778.3  128   
1                 多头  DADA   2018-7-3        236.4  2018-7-5       232.1  423   
2                 多头  DBTC   2018-7-3       6505.5  2018-7-5      6526.5   15   

   trade_costs  net_gains  cum_gains    returns cum_returns  
0       248.61    -607.01    -607.01   -782.73%    -782.73%  
1       247.72   -2066.62   -2673.63  -8742.04%   -9524.78%  
2        73.31     241.69   -2431.93     37.15%   -9487.62%

## 分析大涨大跌

In [11]:
date_fmt = "%Y-%m-%d"
perf_['exit_dt'] = perf_['exit_date'].apply(lambda x: datetime.strptime(x, date_fmt))
perf_['entry_dt'] = perf_['entry_date'].apply(lambda x: datetime.strptime(x, date_fmt))

perf_['holding_days'] = ' '
for i, item in enumerate(perf_['holding_days']):
    perf_['holding_days'][i] = perf_['exit_dt'][i] - perf_['entry_dt'][i] 
    perf_['holding_days'][i] =  perf_['holding_days'][i].days
    
perf_['returns'] = perf_['returns'].apply(lambda x: float(x.replace("%", "")))
perf_['returns'] = perf_['returns']/100

perf_['cum_returns'] = perf_['cum_returns'].apply(lambda x: float(x.replace("%", "")))
perf_['cum_returns'] = perf_['cum_returns']/100

perf_['max_dd'] = perf_.groupby(['asset'])['returns'].min()

perf_[['asset', 'entry_date', 'exit_date', 'returns']].groupby(['asset']).apply(lambda x: x.sort_values('returns').head(3))

C:\Users\Brad Sun\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys
C:\Users\Brad Sun\Anaconda3\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


asset  entry_date   exit_date    returns
asset                                             
DADA  157  DADA   2018-10-7  2018-10-11 -1077.5404
      64   DADA    2018-8-6   2018-8-13  -819.1909
      73   DADA   2018-8-14   2018-8-18  -282.2055
DBCH  68   DBCH   2018-8-11   2018-8-15  -996.9083
      90   DBCH   2018-8-25    2018-9-1  -907.1990
      96   DBCH    2018-9-1    2018-9-6  -871.5163
DBTC  95   DBTC    2018-9-3    2018-9-6   -13.9259
      20   DBTC   2018-7-11   2018-7-16    -8.3207
      44   DBTC   2018-7-29   2018-7-31    -7.2707
DEOS  121  DEOS   2018-9-18   2018-9-20 -1423.1676
      164  DEOS   2018-10-9  2018-10-14 -1054.9849
      55   DEOS    2018-8-3    2018-8-8 -1024.7327
DETH  137  DETH   2018-9-26   2018-9-30  -213.0420
      61   DETH    2018-8-8   2018-8-11  -184.1342
      120  DETH   2018-9-19   2018-9-20  -183.4352
DLTC  106  DLTC    2018-9-1   2018-9-13 -1104.8970
      28   DLTC   2018-7-21   2018-7-23  -604.6588
      125  DLTC   2018-9-19   2018-9-24  -538.6362
DTRX  72   DTRX   2018-8-13   2018-8-17  -491.1295
      162  DTRX  2018-10-11  2018-10-13  -350.3801
      107  DTRX   2018-9-12   2018-9-14  -323.0721
DXRP  102  DXRP    2018-9-6   2018-9-10  -169.4477
      203  DXRP   2018-11-3   2018-11-5  -141.1628
      83   DXRP   2018-8-24   2018-8-27   -43.8984

In [12]:
perf_[['asset', 'entry_date', 'exit_date', 'returns']].groupby(['asset']).apply(lambda x: x.sort_values('returns').tail(3))

asset entry_date   exit_date    returns
asset                                            
DADA  93   DADA   2018-9-1    2018-9-5  1022.3134
      129  DADA  2018-9-18   2018-9-26  1242.4557
      112  DADA   2018-9-9   2018-9-17  1377.5017
DBCH  118  DBCH  2018-9-11   2018-9-19   436.6523
      19   DBCH  2018-7-12   2018-7-16   727.9701
      205  DBCH  2018-11-1   2018-11-5  4015.9346
DBTC  85   DBTC  2018-8-21   2018-8-28    14.0194
      34   DBTC  2018-7-17   2018-7-27    15.3773
      57   DBTC   2018-8-2    2018-8-9    17.2119
DEOS  111  DEOS   2018-9-9   2018-9-16   815.0864
      92   DEOS  2018-8-28    2018-9-3   838.9161
      67   DEOS   2018-8-8   2018-8-14  2038.6107
DETH  91   DETH  2018-8-18    2018-9-3   265.7507
      158  DETH  2018-10-9  2018-10-11   348.6768
      105  DETH   2018-9-5   2018-9-12   443.2830
DLTC  135  DLTC  2018-9-24   2018-9-30   719.6684
      78   DLTC  2018-8-17   2018-8-21   726.0853
      22   DLTC  2018-7-15   2018-7-20   782.7259
DTRX  94   DTRX   2018-9-3    2018-9-5   247.2077
      123  DTRX  2018-9-18   2018-9-24   340.7868
      88   DTRX  2018-8-25    2018-9-1   427.4652
DXRP  36   DXRP  2018-7-14   2018-7-27   291.0797
      66   DXRP   2018-8-7   2018-8-14   393.5054
      128  DXRP  2018-9-17   2018-9-25  2112.6744

In [109]:
# A function that analyzes the performance table from TB regarding which assets gain or lose most:
# Will finish later:

# Daily Price Analysis:
为了更清晰地看清楚每天资产价格的变化幅度，有必要调用价格数据(回头来总结：这部分这个notebook里用不到)

In [118]:
# Pull daily data from SQL data base for all 8 assets:
# asset string format to use: xbtusd_1d
def all_assts_from_sql(asset1, asset2, asset3, asset4, asset5, asset6, asset7, asset8, sql_limit_num):
    conn = psycopg2.connect(database="bitmexdata", user="postgres", password="tongKen123", host="128.199.97.202", port="5432")
    asset1 = "SELECT time, open, high, low, close, volume from {} order by id desc limit {}".format(asset1, sql_limit_num)
    asset2 = "SELECT time, open, high, low, close, volume from {} order by id desc limit {}".format(asset2, sql_limit_num)
    asset3 = "SELECT time, open, high, low, close, volume from {} order by id desc limit {}".format(asset3, sql_limit_num)
    asset4 = "SELECT time, open, high, low, close, volume from {} order by id desc limit {}".format(asset4, sql_limit_num)
    asset5 = "SELECT time, open, high, low, close, volume from {} order by id desc limit {}".format(asset5, sql_limit_num)
    asset6 = "SELECT time, open, high, low, close, volume from {} order by id desc limit {}".format(asset6, sql_limit_num)
    asset7 = "SELECT time, open, high, low, close, volume from {} order by id desc limit {}".format(asset7, sql_limit_num)
    asset8 = "SELECT time, open, high, low, close, volume from {} order by id desc limit {}".format(asset8, sql_limit_num)
    df1 = pd.read_sql(asset1, con=conn)
    df2 = pd.read_sql(asset2, con=conn)
    df3 = pd.read_sql(asset3, con=conn)
    df4 = pd.read_sql(asset4, con=conn)
    df5 = pd.read_sql(asset5, con=conn)
    df6 = pd.read_sql(asset6, con=conn)
    df7 = pd.read_sql(asset7, con=conn)
    df8 = pd.read_sql(asset8, con=conn)
    conn.close()
    return df1, df2, df3, df4, df5, df6, df7, df8

In [ ]:
ada, bch, eos, eth, ltc, trx, btc, xrp = all_assts_from_sql("adau18z18_1d", "bchu18z18_1d", "eosu18z18_1d", 
                                                           "ethu18z18_1d", "ltcu18z18_1d", "trxu18z18_1d", 
                                                           "xbtusd_1d", "xrpu18z18_1d", 200)

# 直接导入下面日线数据

In [15]:
root_2 = "C:/Users/workspace/brad_public_workspace_on_win/SH_tongliang/data/bitmex日线至1106_11_7_2018/bitmex日线至1106/standardized_version/11_7_日线导出数据/导出数据/"
file1 = "DADA(日线).csv"
file2 = "DBCH(日线).csv"
file3 = "DBTC(日线).csv"
file4 = "DEOS(日线).csv"
file5 = "DETH(日线).csv"
file6 = "DLTC(日线).csv"
file7 = "DTRX(日线).csv"
file8 = "DXRP(日线).csv"

ada = pd.read_csv(root_2+file1, engine='python')
bch = pd.read_csv(root_2+file2, engine='python')
btc = pd.read_csv(root_2+file3, engine='python')
eos = pd.read_csv(root_2+file4, engine='python')
eth = pd.read_csv(root_2+file5, engine='python')
ltc = pd.read_csv(root_2+file6, engine='python')
trx = pd.read_csv(root_2+file7, engine='python')
xrp = pd.read_csv(root_2+file8, engine='python')

In [16]:
del ada['0']
del bch['0']
del btc['0']
del eos['0']
del eth['0']
del ltc['0']
del trx['0']
del xrp['0']

In [17]:
ada.columns = ['time', 'open', 'high', 'low', 'close', 'volume']
bch.columns = ['time', 'open', 'high', 'low', 'close', 'volume']
btc.columns = ['time', 'open', 'high', 'low', 'close', 'volume']
eos.columns = ['time', 'open', 'high', 'low', 'close', 'volume']
eth.columns = ['time', 'open', 'high', 'low', 'close', 'volume']
ltc.columns = ['time', 'open', 'high', 'low', 'close', 'volume']
trx.columns = ['time', 'open', 'high', 'low', 'close', 'volume']
xrp.columns = ['time', 'open', 'high', 'low', 'close', 'volume']

In [18]:
def dt_process(df):   
    df.index = pd.to_datetime(df['time'])
    return df

In [19]:
ada_ = dt_process(ada.copy())
bch_ = dt_process(bch.copy())
btc_ = dt_process(btc.copy())
eos_ = dt_process(eos.copy())
eth_ = dt_process(eth.copy())
ltc_ = dt_process(ltc.copy())
trx_ = dt_process(trx.copy())
xrp_ = dt_process(xrp.copy())

In [20]:
def calc_close_pct_chg(df):
    df['pct_chg'] = df['close'].pct_change()
    return df

In [21]:
ada__ = calc_close_pct_chg(ada_.copy())
bch__ = calc_close_pct_chg(bch_.copy())
btc__ = calc_close_pct_chg(btc_.copy())
eos__ = calc_close_pct_chg(eos_.copy())
eth__ = calc_close_pct_chg(eth_.copy())
ltc__ = calc_close_pct_chg(ltc_.copy())
trx__ = calc_close_pct_chg(trx_.copy())
xrp__ = calc_close_pct_chg(xrp_.copy())

In [22]:
# The start/end_date format is like "2018/09/22"
def peek(df, start_date, end_date):
    df_ = df[(df['time']>=start_date) & (df['time'] <= end_date)]
    print (start_date, "to", end_date)
    return df_

## 回撤显著日期一：9/24/2018

In [23]:
# what we lost:
print ("ada 跌幅（期间持仓做多，产生亏损）：")
print(peek(ada__, '2018/09/24', '2018/09/24')['pct_chg'].values)
print ("trx 跌幅（期间持仓做多，产生亏损）：")
print (peek(trx__, '2018/09/24', '2018/09/24')['pct_chg'].values)
print ("xrp 跌幅（期间持仓做多，产生亏损）：")
print (peek(xrp__, '2018/09/24', '2018/09/24')['pct_chg'].values)
# What we gained
print ("====================================================")
print ("btc 跌幅（期间持仓做空，产生盈利）：")
print(peek(btc__, '2018/09/24', '2018/09/24')['pct_chg'].values)
print ("ltc 跌幅（期间持仓做空，产生盈利）：")
print (peek(ltc__, '2018/09/24', '2018/09/24')['pct_chg'].values)
print ("eos 跌幅（期间持仓做空，产生盈利）：")
print (peek(eos__, '2018/09/24', '2018/09/24')['pct_chg'].values)

ada 跌幅（期间持仓做多，产生亏损）：
2018/09/24 to 2018/09/24
[-0.0751634]
trx 跌幅（期间持仓做多，产生亏损）：
2018/09/24 to 2018/09/24
[-0.07482993]
xrp 跌幅（期间持仓做多，产生亏损）：
2018/09/24 to 2018/09/24
[-0.13030286]
btc 跌幅（期间持仓做空，产生盈利）：
2018/09/24 to 2018/09/24
[-0.01857377]
ltc 跌幅（期间持仓做空，产生盈利）：
2018/09/24 to 2018/09/24
[-0.03810561]
eos 跌幅（期间持仓做空，产生盈利）：
2018/09/24 to 2018/09/24
[-0.04579985]


In [25]:
def take_pct_chg_from_df(df):
    df = df['pct_chg']
    return df

In [26]:
take_pct_chg_from_df(peek(ada__, '2018/07/26', '2018/08/01'))

2018/07/26 to 2018/08/01


time
2018-07-26   -0.024646
2018-07-27   -0.016846
2018-07-28   -0.016659
2018-07-29    0.002420
2018-07-30   -0.066634
2018-07-31   -0.039317
2018-08-01    0.017232
Name: pct_chg, dtype: float64

In [354]:
print (peek(ada__, '2018/08/01', '2018/08/01')['pct_chg'].values)

2018/08/01 to 2018/08/01
[0.01723209]


In [31]:
print (peek(trx__, '2018/10/11', '2018/10/11')['pct_chg'].values)

2018/10/11 to 2018/10/11
[-0.1511335]


## 回撤显著日期二：10/11/2018

In [213]:
print (peek(trx__, '2018/10/11', '2018/10/11')['pct_chg'].values)

2018/10/11 to 2018/10/11
[-0.1511335]


# 到目前，以观察出问题最大的两个资产：ADA和TRX

ADA在持仓到第七天，于9/24大跌7.5%，大幅冲击做多仓位；

TRX再持仓到第八天，于10/11大跌15.11%，大幅冲击做多仓位。

假设：既然如此，是否可以从持仓时长入手？

In [228]:
# 首先要设一个标准，按照这个标准去比较持仓长和持仓段到底平均来说哪个更赚钱？
# 假设这个量化标准为 （净利润/持仓天数） = 每天平均利润
# 在此之前要把 holding_days = 0的资产剔除

In [231]:
perf__ = perf_.copy()
perf_no_zero_hdng_d = perf__[perf__['holding_days']!=0]

In [234]:
perf_no_zero_hdng_d['net_gains_per_day'] = perf_no_zero_hdng_d['net_gains']/perf_no_zero_hdng_d['holding_days']

C:\Users\Brad Sun\Anaconda3\lib\site-packages\ipykernel\__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


In [240]:
def avg_gains_per_day(df, holding_days):
    avg_gains_per_day = df[df['holding_days']==holding_days]['net_gains_per_day'].mean()
    std_gains_per_day = df[df['holding_days']==holding_days]['net_gains_per_day'].std()
    net_gains_per_day = df[df['holding_days']==holding_days]['net_gains_per_day']
    print ("Holding Days: ", holding_days)
    print ("Avg Daily Gains: ", avg_gains_per_day)
    print ("Stddev of Daily Gains", std_gains_per_day)
    return net_gains_per_day

In [241]:
results_1 = avg_gains_per_day(perf_no_zero_hdng_d, 1)
results_2 = avg_gains_per_day(perf_no_zero_hdng_d, 2)
results_3 = avg_gains_per_day(perf_no_zero_hdng_d, 3)
results_4 = avg_gains_per_day(perf_no_zero_hdng_d, 4)
results_5 = avg_gains_per_day(perf_no_zero_hdng_d, 5)
results_6 = avg_gains_per_day(perf_no_zero_hdng_d, 6)
results_7 = avg_gains_per_day(perf_no_zero_hdng_d, 7)
results_8 = avg_gains_per_day(perf_no_zero_hdng_d, 8)
results_9 = avg_gains_per_day(perf_no_zero_hdng_d, 9)
results_10 = avg_gains_per_day(perf_no_zero_hdng_d, 10)
results_11 = avg_gains_per_day(perf_no_zero_hdng_d, 11)
results_12 = avg_gains_per_day(perf_no_zero_hdng_d, 12)

Holding Days:  1
Avg Daily Gains:  -479.6520408163265
Stddev of Daily Gains 2677.011495622637
Holding Days:  2
Avg Daily Gains:  -597.2528378378379
Stddev of Daily Gains 2690.4629154386703
Holding Days:  3
Avg Daily Gains:  -379.93410256410255
Stddev of Daily Gains 1287.4611901408848
Holding Days:  4
Avg Daily Gains:  -70.99858695652189
Stddev of Daily Gains 2294.323526507723
Holding Days:  5
Avg Daily Gains:  -378.5047368421052
Stddev of Daily Gains 1229.9011769379795
Holding Days:  6
Avg Daily Gains:  868.5069791666666
Stddev of Daily Gains 918.0939586986561
Holding Days:  7
Avg Daily Gains:  755.0562857142857
Stddev of Daily Gains 1687.7305284258473
Holding Days:  8
Avg Daily Gains:  2457.0575
Stddev of Daily Gains 4020.379789097848
Holding Days:  9
Avg Daily Gains:  674.2733333333333
Stddev of Daily Gains 275.10067889482764
Holding Days:  10
Avg Daily Gains:  553.6985
Stddev of Daily Gains 808.4938727934182
Holding Days:  11
Avg Daily Gains:  1497.0204545454546
Stddev of Daily Gain

In [236]:
perf_no_zero_hdng_d[perf_no_zero_hdng_d['holding_days']==1]['net_gains_per_day']

0      -607.01
3     -1179.91
4     -2398.69
12    -2718.97
16     -291.44
25     1255.07
27     6288.34
29     -775.73
31     1418.33
38      332.95
41     -590.46
42       51.98
58    -1626.17
59      575.25
60    -4504.45
62    -1722.27
63     -530.25
65     9593.73
86    -2508.67
98     2749.03
101    -249.98
104    6605.17
108    -1106.9
110     210.74
117   -4310.35
120   -6033.18
126     817.16
127     1983.8
138   -1376.44
139    -661.12
145     -860.1
146   -2597.24
149   -2038.95
152     472.26
154   -1605.59
156     -29.09
161   -2397.57
166   -2044.16
168    -155.83
171   -1880.08
172     170.68
173    -249.48
180     -64.64
182    -755.69
183   -1840.14
188    -493.99
192   -4851.24
196       44.9
201   -1016.56
Name: net_gains_per_day, dtype: object

In [348]:
perf_[perf_['holding_days']>=5]

position_direction asset  entry_date  entry_price   exit_date  exit_price  \
7                   多头  DEOS    2018-7-3       136.47    2018-7-8      131.26   
9                   空头  DTRX    2018-7-4       614.00   2018-7-10      540.00   
11                  多头  DBTC    2018-7-6      6599.50   2018-7-11     6371.00   
13                  多头  DETH    2018-7-6       715.10   2018-7-12      694.00   
15                  空头  DADA    2018-7-6       224.90   2018-7-13      225.00   
17                  空头  DEOS    2018-7-9       116.87   2018-7-15      117.29   
20                  空头  DBTC   2018-7-11      6371.00   2018-7-16     6719.50   
22                  空头  DLTC   2018-7-15       125.80   2018-7-20      113.10   
23                  空头  DETH   2018-7-16       721.40   2018-7-21      632.20   
24                  多头  DEOS   2018-7-16       122.47   2018-7-21      109.75   
26                  多头  DADA   2018-7-13       225.00   2018-7-22      235.30   
33                  空头  DTRX   2018-7-20       486.00   2018-7-26      467.00   
34                  多头  DBTC   2018-7-17      7318.50   2018-7-27     8190.00   
36                  空头  DXRP   2018-7-14       726.60   2018-7-27      570.60   
48                  多头  DBCH   2018-7-29       104.40    2018-8-3      100.00   
50                  多头  DXRP   2018-7-31       574.60    2018-8-6      604.20   
51                  空头  DADA   2018-7-26       213.70    2018-8-6      188.50   
53                  空头  DTRX    2018-8-1       435.00    2018-8-7      416.00   
55                  多头  DEOS    2018-8-3        99.31    2018-8-8       89.43   
57                  空头  DBTC    2018-8-2      7539.50    2018-8-9     6536.00   
64                  多头  DADA    2018-8-6       188.50   2018-8-13      159.80   
66                  空头  DXRP    2018-8-7       573.20   2018-8-14      442.30   
67                  空头  DEOS    2018-8-8        89.43   2018-8-14       72.92   
71                  多头  DBTC   2018-8-11      6241.00   2018-8-17     6578.50   
76                  多头  DEOS   2018-8-15        73.02   2018-8-21       75.39   
80                  多头  DXRP   2018-8-16       461.20   2018-8-24      489.60   
81                  空头  DADA   2018-8-20       146.20   2018-8-25      142.00   
85                  多头  DBTC   2018-8-21      6467.00   2018-8-28     7076.50   
88                  多头  DTRX   2018-8-25       328.00    2018-9-1      375.00   
90                  空头  DBCH   2018-8-25        80.70    2018-9-1       86.40   
..                 ...   ...         ...          ...         ...         ...   
92                  多头  DEOS   2018-8-28        83.31    2018-9-3       89.35   
96                  多头  DBCH    2018-9-1        86.40    2018-9-6       80.10   
97                  空头  DXRP   2018-8-28       504.00    2018-9-6      463.30   
105                 空头  DETH    2018-9-5       341.30   2018-9-12      288.70   
106                 多头  DLTC    2018-9-1        92.80   2018-9-13       83.50   
109                 多头  DBTC   2018-9-10      6300.00   2018-9-15     6515.50   
111                 多头  DEOS    2018-9-9        78.30   2018-9-16       83.50   
112                 空头  DADA    2018-9-9       122.30   2018-9-17      101.40   
118                 空头  DBCH   2018-9-11        69.80   2018-9-19       67.50   
123                 多头  DTRX   2018-9-18       307.00   2018-9-24      340.00   
125                 空头  DLTC   2018-9-19        84.70   2018-9-24       88.35   
128                 多头  DXRP   2018-9-17       433.20   2018-9-25      832.70   
129                 多头  DADA   2018-9-18       109.30   2018-9-26      124.45   
134                 空头  DTRX   2018-9-25       336.00   2018-9-30      347.00   
135                 多头  DLTC   2018-9-24        88.35   2018-9-30       94.20   
143                 多头  DXRP   2018-9-27       826.70   2018-10-3      822.80   
159                 多头  DTRX   2018-10-4       351.00  2018-10-11      337.00   
164                 多头  DEOS   2018

# 发现和假设不一样，事实持仓时间更长反而平均盈利更多，然而有新的启发和发现：

问题在于应该将排名位置也考虑进去：

加入排名限制，得出新假设：排名连续N天未发生变化并且已经N天持仓，即需要平仓，N = 5 or N = 6

In [242]:
import numpy as np
import pandas as pd
import datetime
import time
import re
import talib
import operator
import psycopg2
import smtplib
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from email.mime.image import MIMEImage
from email.header import Header

"""
自动数据库读取数据，计算排序，发送邮件
# py file to refer to: zdsort.py
"""


# 计算两天close涨幅 T1,T2
def zf(data):
    resdata = []
    for r in data.itertuples():
        time = getattr(r, 'time')
        open = getattr(r, 'open')
        high = getattr(r, 'high')
        low = getattr(r, 'low')
        close = getattr(r, 'close')
        resdata.append([time, open, high, low, close])
    res = [0]
    for i in range(1, len(resdata)):
        T = (float(resdata[i][4]) - float(resdata[i - 1][4])) / float(resdata[i - 1][4]) * 100
        res.append(T)
    return res


# 计算TP和值
def myTP(T1, T2, Q):
    res = [0]

    for i in range(len(T1)):
        t = T1[i] * (1 - Q / 100) + T2[i] * (Q / 100)
        res.append(t)
    return res


# 计算TP差值
def myTM(T1, T2, Q):
    res = [0]
    for i in range(len(T1)):
        t = T1[i] * (1 - Q / 100) - T2[i] * (Q / 100)
        res.append(t)
    return res


# 计算和值累加 或 差值累加
def mySUM(TP):
    res = [0]
    for i in TP:
        res.append(res[-1] + i)
    return res


# 求指定周期均值，参数1数据列表，参数2周期，返回单个数值
def myAverage(lists, tp):
    if len(lists) > tp:
        return np.mean(lists[-tp:])
    else:
        return np.mean(lists)


# 判断趋势
def pdqs(lists):
    count = -1
    while lists[count] == lists[count - 1]:
        count -= 1
    if lists[count] > lists[count - 1]:
        return "趋势向上"
    elif lists[count] < lists[count - 1]:
        return "趋势向下"


# 判断交叉
# def pdjx(ddlist, eelist):
#     for i in range(len(ddlist)):
#         if ddlist[-i]:
#             return "金叉"
#         elif eelist[-i]:
#             return "死叉"
#         else:
#             return '没有交叉'
# 判断交叉
def pdjx(ddlist, eelist):
    for i in range(len(ddlist)):
        if ddlist[-i]:
            return "金叉"
        elif eelist[-i]:
            return "死叉"
        else:
            return '没有交叉'


# 计算结果 参数：数据1，数据2，短周期，长周期，权重1，权重2,截止日期
def comres(data0, data1, S, L, Q1, Q2, starttime, endtime):
    """计算两者排名"""
    pd.to_datetime(data0.ix[:, 0])  # 转换为时间格式
    pd.to_datetime(data1.ix[:, 0])  # 转换为时间格式

    # timestr = '2018-10-19 00:00:00'

    data0 = data0[(data0.ix[:, 0] <= endtime) & (data0.ix[:, 0] >= starttime)]
    data1 = data1[(data1.ix[:, 0] <= endtime) & (data1.ix[:, 0] >= starttime)]
    print('筛选后')
    print(len(data0), len(data1))
    T1 = zf(data0)  # 第一腿涨幅
    T2 = zf(data1)  # 第二腿涨幅

    TP = myTP(T1, T2, Q1)  # 相对和
    TM = myTM(T1, T2, Q2)  # 相对差

    TPSUM = mySUM(TP)  # 相对和累加
    TMSUM = mySUM(TM)  # 相对差累加

    TMS = np.array(TMSUM)
    TPS = np.array(TPSUM)

    MA1 = talib.MA(TMS, S)  # 相对差短周期均线
    MA2 = talib.MA(TMS, L)  # 相对差长周期均线

    # MAS = talib.MA(TPS, S)
    # MAL = talib.MA(TPS, L)

    if MA1[-1] > MA2[-1]:
        jx = '金叉'
    else:
        jx = '死叉'

    MA1qs = pdqs(MA1)
    MA2qs = pdqs(MA2)
    # print(MA2)
    # print("交叉：" + jx)
    # print("短周期相对差累加均线MA1趋势:" + MA1qs)
    # print("长周期相对差累加均线MA2趋势:" + MA2qs)
    if jx == '金叉':
        return 1
    if jx == '死叉':
        return -1


def get_data(conn, sym, period):
    """获取数据"""
    if sym == 'BTC':
        sym = 'xbtusd'
    else:
        sym = sym.lower() + 'u18z18'
    sqllasttime = " SELECT time, open, high, low, close, volume from " + sym + "_" + period + " order by time"
    df = pd.read_sql(sqllasttime, con=conn)
    return df

ImportError: No module named 'talib'

In [340]:
root_3 = "C:/Users/workspace/SH_tongliang/"
rank = "11_8_rank_rest.csv"
rank = pd.read_csv(root_3+rank, encoding="latin-1", delimiter=',')

In [343]:
df = pd.DataFrame(rank.stack()).reset_index()
del df['level_0']
df.columns = ['date', 'to_split']
df['asset'], df['rank'] = df['to_split'].str.split(' ', 1).str
del df['to_split']
df['rank'] = df['rank'].apply(lambda x: int(x))
pd.pivot_table(df, values='rank', columns='asset', index = 'date')

asset      ADA  BCH  BTC  EOS  ETH  LTC  TRX  XRP
date                                             
2018-11-7    5    7    0    2    4    1    3    6
2018-11-8    3    6    2    5    4    0    1    7

In [345]:
def export_formmated_rank(location, rank_file):
    location = location
    rank_file = rank_file
    rank = pd.read_csv(location + rank_file, encoding="latin-1", delimiter=',')
    df = pd.DataFrame(rank.stack()).reset_index()
    del df['level_0']
    df.columns = ['date', 'to_split']
    df['asset'], df['rank'] = df['to_split'].str.split(' ', 1).str
    del df['to_split']
    df['rank'] = df['rank'].apply(lambda x: int(x))
    formatted_df = pd.pivot_table(df, values='rank', columns='asset', index = 'date')
    return formatted_df

# Build performance analysis functions 

In [ ]:
# Break down code:

location = "F:/同梁智能科技/projects/策略库/"
perf = pd.read_csv(location+"11_8_Ranking_Original_Half_Commissions_3_5_50_drawdown_analysis_csv.csv")

In [ ]:
perf_.columns = ['position_direction', 'asset', 'entry_date', 'entry_price', 'exit_date', 'exit_price', 'qty', 'trade_costs', 'net_gains', 'cum_gains',
       'returns', 'cum_returns']

In [357]:
# Import raw performance csv file and build into a data table in a ready-to-process format:
def build_perf_df(location, perf_file):
    perf1 = pd.read_csv(location+perf_file)
    del perf1['Unnamed: 0']
    perf1.columns = ['position_direction', 'asset', 'entry_date', 'entry_price', 
                    'exit_date', 'exit_price', 'qty', 'trade_costs', 'net_gains', 
                    'cum_gains', 'returns', 'cum_returns']
    return perf1    

In [359]:
# Process performance table so that it can be used to generate further asset perf analysis:
def process_perf_df(perf1):
    date_fmt = "%Y-%m-%d"
    perf1['exit_dt'] = perf1['exit_date'].apply(lambda x: datetime.strptime(x, date_fmt))
    perf1['entry_dt'] = perf1['entry_date'].apply(lambda x: datetime.strptime(x, date_fmt))

    perf1['holding_days'] = ' '
    for i, item in enumerate(perf1['holding_days']):
        perf1['holding_days'][i] = perf1['exit_dt'][i] - perf1['entry_dt'][i] 
        perf1['holding_days'][i] =  perf1['holding_days'][i].days

    perf1['returns'] = perf1['returns'].apply(lambda x: float(x.replace("%", "")))
    perf1['returns'] = perf1['returns']/100

    perf1['cum_returns'] = perf1['cum_returns'].apply(lambda x: float(x.replace("%", "")))
    perf1['cum_returns'] = perf1['cum_returns']/100
    
    perf2 = perf1.copy()
    return perf2

In [388]:
def perf_df_analysis(perf2, show_minmax_returns = True, show_return_rank = True):
    perf2['worst_return'] = perf2.groupby(['asset'])['returns'].min()
    perf2['best_return'] = perf2.groupby(['asset'])['returns'].max()
    perf2_rank = perf2.copy()[['asset', 'entry_date', 'exit_date', 'returns']].groupby(['asset']).apply(lambda x: x.sort_values('returns'))
    if show_minmax_returns == True and show_return_rank == True:
                                                                                                        
        print ("worst_return: ", perf2['worst_return'])
        print ("best return: ", perf2['best_return'])
        print ("return ranks", perf2_rank)                                                                                     
    if show_minmax_returns == True and show_return_rank == False:
                                                                                                        
        print ("worst_return: ", perf2['worst_return'])
        print ("best return: ", perf2['best_return'])
    if show_minmax_returns == False and show_return_rank == True:
                                                                                                    
        print ("return ranks: ", perf2_rank)
    else:
        return 